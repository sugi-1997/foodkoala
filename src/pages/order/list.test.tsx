/**
 * @jest-environment jsdom
 */

import Orderlist from './list';
import { render, screen, waitFor } from '@testing-library/react';
import useSWR, { Middleware, SWRConfig, SWRResponse } from 'swr';
import { useRouter } from 'next/router';
import mockRouter from 'next-router-mock';

jest.mock('next/router', () => ({
  useRouter() {
    return {
      asPath: '/login',
    };
  },
}));

const fetchMockError = () => {
  const error = new Error('Failed to fetch data');
  return new Promise((rejects) => {
    rejects({
      ok: false,
      status: 400,
      json: async () => {
        throw error;
      },
    });
  });
};

describe('OrderList', () => {
  beforeAll(() => {
    (global as any).fetch = jest.fn();
    jest.mock('next/router', () => require('next-router-mock'));
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  it('renders loding when data is not fetched', () => {
    render(<Orderlist />);
    expect(screen.getByText('Loading...')).toBeInTheDocument();
  });

  it('renders Error when fetching data is failed', async () => {
    global.fetch = jest.fn().mockImplementation(fetchMockError);

    const { asFragment } = render(<Orderlist />);

    expect(await screen.findByText('Error...')).toBeInTheDocument();
    expect(fetch).toHaveBeenCalledWith('/api/get_cart_items');
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders OrderList correctly when the cart is empty', async () => {
    jest.mock('swr');
    (global as any).fetch.mockResolvedValue({
      ok: true,
      status: 200,
      json: async () => [],
    });

    // Orderlistをレンダリング
    const { asFragment } = render(
      <SWRConfig>
        <Orderlist />
      </SWRConfig>
    );

    //エラーメッセージが画面に描写されているか確認
    const errorMessage = await screen.findByText(
      '※カートに商品がないため、購入できません'
    );
    await waitFor(() => expect(errorMessage).toBeInTheDocument());

    //リンクが画面に描写されているか確認
    const link = screen.getByRole('link', { name: 'メニュー一覧へ' });
    await waitFor(() => expect(link).toBeInTheDocument());

    //fetchの引数が正しいか確認
    expect(fetch).toHaveBeenCalledWith('/api/get_cart_items');

    //スナップショットテスト
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders Orderlist when items in the cart', async () => {
    const mockData = [
      {
        id: 1,
        cart_id: 1,
        item_id: 5,
        count: 1,
      },
    ];
    jest.mock('swr');
    (global as any).fetch.mockResolvedValue({
      ok: true,
      status: 200,
      json: async () => mockData,
    });

    const { asFragment } = render(
      <SWRConfig>
        <Orderlist />
      </SWRConfig>
    );
    // 画面に商品が描写されているかの確認

    expect(await screen.findByText('焼きいも')).toBeInTheDocument();

    // fetchの引数が正しいかの確認
    expect(fetch).toHaveBeenCalledWith('/api/get_cart_items');
    // スナップショットテストの実施
    expect(asFragment()).toMatchSnapshot();
  });
});
