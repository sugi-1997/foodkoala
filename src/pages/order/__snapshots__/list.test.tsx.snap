// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderList renders Error when fetching data is failed 1`] = `
<DocumentFragment>
  <div>
    Error...
  </div>
</DocumentFragment>
`;

exports[`OrderList renders OrderList correctly when the cart is empty 1`] = `
<DocumentFragment>
  <header
    class="header"
  >
    <div
      class="logo"
    >
      <a
        href="/"
      >
        <img
          alt="logo"
          data-nimg="1"
          decoding="async"
          height="35"
          loading="lazy"
          src="/_next/image?url=%2Fimages%2Fprovisional_logo.png&w=96&q=75"
          srcset="/_next/image?url=%2Fimages%2Fprovisional_logo.png&w=48&q=75 1x, /_next/image?url=%2Fimages%2Fprovisional_logo.png&w=96&q=75 2x"
          style="color: transparent;"
          width="35"
        />
        <span>
          Food Koala
        </span>
      </a>
    </div>
    <div
      class="hamburger"
    >
      <input
        class="menu_btn_check"
        id="menu_btn_check"
        type="checkbox"
      />
      <label
        class="menu_btn"
        for="menu_btn_check"
      >
        <span />
      </label>
      <nav
        class="nav"
      >
        <ul>
          <li>
            <a
              href="/"
            >
              メニュー
            </a>
          </li>
          <li>
            <a
              href="/shop/list"
            >
              ショップ
            </a>
          </li>
          <li>
            <a
              href="/shop/favorite"
            >
              お気に入り
            </a>
          </li>
          <li>
            <a
              href="/order/list"
            >
              注文リスト
            </a>
          </li>
          <li>
            <a
              href="/order/order_history"
            >
              注文履歴
            </a>
          </li>
          <li
            class="true"
          >
            <a
              href="/login"
            >
              ログイン
            </a>
          </li>
          <li
            class="false"
          >
            ログアウト
          </li>
        </ul>
      </nav>
    </div>
  </header>
  <div>
    <ul
      class="breadList"
    >
      <li>
        <a
          href="/"
        >
          Food Koala トップ
        </a>
      </li>
      <li>
        <i
          class="fa-solid fa-forward"
        />
      </li>
      <li>
        <a
          href="/order/list"
        >
          注文リスト
        </a>
      </li>
      <li>
        <i
          class="fa-solid fa-forward"
        />
      </li>
    </ul>
  </div>
  <div
    class="order_list"
  >
    <div>
      <h1>
        注文リスト
      </h1>
    </div>
    <div
      class="order_list"
    >
      <br />
      <div
        class="order_list_sub"
      >
        カートに商品はありません
      </div>
      <br />
    </div>
    <div
      class="alert"
    >
      <p>
        ※カートに商品がないため、購入できません
      </p>
      <button>
        <a
          href="/"
        >
          メニュー一覧へ
        </a>
      </button>
    </div>
  </div>
  <footer
    class="footer"
  >
    <a
      class="link"
      href="/concept"
    >
      About us
    </a>
    <a
      class="link"
      href="/inquiry_form"
    >
      お問い合わせ
    </a>
  </footer>
</DocumentFragment>
`;

exports[`OrderList renders Orderlist when items in the cart 1`] = `
<DocumentFragment>
  <header
    class="header"
  >
    <div
      class="logo"
    >
      <a
        href="/"
      >
        <img
          alt="logo"
          data-nimg="1"
          decoding="async"
          height="35"
          loading="lazy"
          src="/_next/image?url=%2Fimages%2Fprovisional_logo.png&w=96&q=75"
          srcset="/_next/image?url=%2Fimages%2Fprovisional_logo.png&w=48&q=75 1x, /_next/image?url=%2Fimages%2Fprovisional_logo.png&w=96&q=75 2x"
          style="color: transparent;"
          width="35"
        />
        <span>
          Food Koala
        </span>
      </a>
    </div>
    <div
      class="hamburger"
    >
      <input
        class="menu_btn_check"
        id="menu_btn_check"
        type="checkbox"
      />
      <label
        class="menu_btn"
        for="menu_btn_check"
      >
        <span />
      </label>
      <nav
        class="nav"
      >
        <ul>
          <li>
            <a
              href="/"
            >
              メニュー
            </a>
          </li>
          <li>
            <a
              href="/shop/list"
            >
              ショップ
            </a>
          </li>
          <li>
            <a
              href="/shop/favorite"
            >
              お気に入り
            </a>
          </li>
          <li>
            <a
              href="/order/list"
            >
              注文リスト
            </a>
          </li>
          <li>
            <a
              href="/order/order_history"
            >
              注文履歴
            </a>
          </li>
          <li
            class="true"
          >
            <a
              href="/login"
            >
              ログイン
            </a>
          </li>
          <li
            class="false"
          >
            ログアウト
          </li>
        </ul>
      </nav>
    </div>
  </header>
  <div>
    <ul
      class="breadList"
    >
      <li>
        <a
          href="/"
        >
          Food Koala トップ
        </a>
      </li>
      <li>
        <i
          class="fa-solid fa-forward"
        />
      </li>
      <li>
        <a
          href="/order/list"
        >
          注文リスト
        </a>
      </li>
      <li>
        <i
          class="fa-solid fa-forward"
        />
      </li>
    </ul>
  </div>
  <div
    class="order_list"
  >
    <div>
      <h1>
        注文リスト
      </h1>
    </div>
    <div
      class="order_list"
    >
      <br />
      <div
        class="order_list_sub"
      >
        カートに商品はありません
      </div>
      <br />
    </div>
    <div
      class="alert"
    >
      <p>
        ※カートに商品がないため、購入できません
      </p>
      <button>
        <a
          href="/"
        >
          メニュー一覧へ
        </a>
      </button>
    </div>
  </div>
  <footer
    class="footer"
  >
    <a
      class="link"
      href="/concept"
    >
      About us
    </a>
    <a
      class="link"
      href="/inquiry_form"
    >
      お問い合わせ
    </a>
  </footer>
</DocumentFragment>
`;

exports[`renders Orderlist when items in the cart 1`] = `
<DocumentFragment>
  <header
    class="header"
  >
    <div
      class="logo"
    >
      <a
        href="/"
      >
        <img
          alt="logo"
          data-nimg="1"
          decoding="async"
          height="35"
          loading="lazy"
          src="/_next/image?url=%2Fimages%2Fprovisional_logo.png&w=96&q=75"
          srcset="/_next/image?url=%2Fimages%2Fprovisional_logo.png&w=48&q=75 1x, /_next/image?url=%2Fimages%2Fprovisional_logo.png&w=96&q=75 2x"
          style="color: transparent;"
          width="35"
        />
        <span>
          Food Koala
        </span>
      </a>
    </div>
    <div
      class="hamburger"
    >
      <input
        class="menu_btn_check"
        id="menu_btn_check"
        type="checkbox"
      />
      <label
        class="menu_btn"
        for="menu_btn_check"
      >
        <span />
      </label>
      <nav
        class="nav"
      >
        <ul>
          <li>
            <a
              href="/"
            >
              メニュー
            </a>
          </li>
          <li>
            <a
              href="/shop/list"
            >
              ショップ
            </a>
          </li>
          <li>
            <a
              href="/shop/favorite"
            >
              お気に入り
            </a>
          </li>
          <li>
            <a
              href="/order/list"
            >
              注文リスト
            </a>
          </li>
          <li>
            <a
              href="/order/order_history"
            >
              注文履歴
            </a>
          </li>
          <li
            class="true"
          >
            <a
              href="/login"
            >
              ログイン
            </a>
          </li>
          <li
            class="false"
          >
            ログアウト
          </li>
        </ul>
      </nav>
    </div>
  </header>
  <div>
    <ul
      class="breadList"
    >
      <li>
        <a
          href="/"
        >
          Food Koala トップ
        </a>
      </li>
      <li>
        <i
          class="fa-solid fa-forward"
        />
      </li>
      <li>
        <a
          href="/order/list"
        >
          注文リスト
        </a>
      </li>
      <li>
        <i
          class="fa-solid fa-forward"
        />
      </li>
    </ul>
  </div>
  <div
    class="order_list"
  >
    <div
      class="h1"
    >
      <h1>
        注文リスト
      </h1>
    </div>
    <div
      class="order_list"
    >
      <div>
        <dl>
          <dt />
          <dd>
            <img
              alt="商品画像"
              data-nimg="1"
              decoding="async"
              height="150"
              loading="lazy"
              src=""
              style="color: transparent;"
              width="150"
            />
          </dd>
          <dd>
            <select
              id="itemCount-0"
              name="itemCount"
            >
              <option
                value="1"
              >
                1
              </option>
              <option
                value="2"
              >
                2
              </option>
              <option
                value="3"
              >
                3
              </option>
            </select>
            個
          </dd>
          <dd>
            NaN円
          </dd>
          <dd>
            <button
              class="delete_button"
              value="1"
            >
              削除
            </button>
          </dd>
        </dl>
        <p>
          小計：NaN円
        </p>
      </div>
    </div>
    <button
      class="transition_button"
    >
      購入画面へ
    </button>
  </div>
  <footer
    class="footer"
  >
    <a
      class="link"
      href="/concept"
    >
      About us
    </a>
    <a
      class="link"
      href="/inquiry_form"
    >
      お問い合わせ
    </a>
  </footer>
</DocumentFragment>
`;
